### PR TITLE
feat(sdk-coin-sol): add raw msg authorize builder

### PR DESCRIPTION
**Pull request by ewangbitgo**
Ticket: EA-1169

Build a transaction from raw msg generated from SOLANA-CLI. 

solana.web3.js cannot fully parse the transaction generated by CLI so some workaround has to be made to extract information.

### CODE REVIEW

**Comment by zahin-mohammad** [comment]
> solana.web3.js cannot fully parse the transaction generated by CLI so some workaround has to be made to extract information.

Can you expand on this? Or give an example tx that can't be parsed.

**Comment by zahin-mohammad** [comment]
> > solana.web3.js cannot fully parse the transaction generated by CLI so some workaround has to be made to extract information.
> 
> Can you expand on this? Or give an example tx that can't be parsed.

NVM i see some examples in the PR, taking a look.

File: modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+  protected validateMessage(msg: string): boolean {
+    const tx = SOLTransaction.populate(SOLMessage.from(Buffer.from(msg, 'base64')), []);
+    const instructions = tx.instructions;
+    assert(instructions.length === 2, 'Invalid transaction, expected 2 instruction, got ' + instructions.length);
+    for (const instruction of instructions) {
+      switch (instruction.programId.toString()) {
+        case SystemProgram.programId.toString():
+          const instructionName = SystemInstruction.decodeInstructionType(instruction);
+          assert(instructionName === 'AdvanceNonceAccount', 'Invalid transaction:' + instructionName);
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts:82]: Please do not use magic strings. Define these as constants somewhere.

File: modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+          break;
+        case StakeProgram.programId.toString():
+          const data = instruction.data.toString('hex');
+          assert(data === '0a00000001000000', 'Invalid staking instruction data:' + data);
+          break;
+        default:
+          assert(
+            false,
+            'Invalid transaction, instruction program id not supported: ' + instruction.programId.toString()
+          );
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts:92]: You can just throw an error, asserting a constant false value is not very intuitive. 

```suggestion
	throw new Error(`Invalid transaction, instruction program id not supported: ${instruction.programId.toString()}`);
```

File: modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    throw new NotSupported('Method not supported on this builder');
+  }
+
+  validateKey(key: BaseKey): void {
+    throw new NotSupported('Method not supported on this builder');
+  }
+
+  validateRawTransaction(rawTransaction: string): void {
+    this.validateRawTransaction(rawTransaction);
+  }
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts:124]: Isn't this just an infinite loop?

File: modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+          assert(
+            false,
+            'Invalid transaction, instruction program id not supported: ' + instruction.programId.toString()
+          );
+      }
+    }
+    return true;
+  }
+
+  protected fromImplementation(rawTransaction: string): BaseTransaction {
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts:98]: Is there anything that tests this function in a unit test?

File: modules/sdk-coin-sol/src/lib/utils.ts [hunk 5]
```diff
@@ -330,6 +333,27 @@ export function validateIntructionTypes(instructions: TransactionInstruction[]):
...
+ */
+export function validateRawMsgInstruction(instructions: TransactionInstruction[]): boolean {
+  // as web3.js cannot decode authorize instruction from CLI, we need to check it manually first
+  if (instructions.length === 2) {
+    const programId1 = instructions[0].programId.toString();
+    const programId2 = instructions[1].programId.toString();
+    if (programId1 === SystemProgram.programId.toString() && programId2 === StakeProgram.programId.toString()) {
+      const instructionName1 = SystemInstruction.decodeInstructionType(instructions[0]);
+      const data = instructions[1].data.toString('hex');
+      if (instructionName1 === 'AdvanceNonceAccount' && data === '0a00000001000000') {
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/lib/utils.ts:350]: Please define these strings somewhere. (or use a value exposed by the library)

File: modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts [hunk 6]
```diff
@@ -0,0 +1,47 @@
+import should from 'should';
+
+import * as testData from '../../resources/sol';
+import { getBuilderFactory } from '../getBuilderFactory';
+import { Utils } from '../../../src';
+
+describe('Sol Staking Raw Message Authorize Builder', () => {
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts:7]: Please add non-happy path tests.

File: modules/sdk-coin-sol/src/sol.ts [hunk 7]
```diff
@@ -415,9 +415,13 @@ export class Sol extends BaseCoin {
     let rebuiltTransaction;
 
     try {
-      const transactionBuilder = factory.from(params.txBase64).fee({ amount: params.feeInfo.fee });
-      if (params.tokenAccountRentExemptAmount) {
-        transactionBuilder.associatedTokenAccountRent(params.tokenAccountRentExemptAmount);
+      const transactionBuilder = factory.from(params.txBase64);
+      if (transactionBuilder instanceof TransactionBuilder) {
+        const txBuilder = transactionBuilder as TransactionBuilder;
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/sol.ts:420]: Do you still need the case here if it's inside the `instanceof` if statement?

File: modules/sdk-coin-sol/src/lib/transaction.ts [hunk 8]
```diff
@@ -461,4 +477,47 @@ export class Transaction extends BaseTransaction {
       durableNonce: durableNonce,
     };
   }
+
+  private explainRawMsgAuthorizeTransaction(): TransactionExplanation {
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/lib/transaction.ts:481]: if i added an extra ix, this fn would not display it. I understand that your other methods would hopefully throw an error upstream if that where the case, but in isolation, it wouldnt catch this issue. Just something to be mindful of. 

Ideally decode/explain functions take into account the entire tx without assuming context of upstream validation.

File: modules/sdk-coin-sol/src/lib/instructionParamsFactory.ts [hunk 9]
```diff
@@ -59,6 +59,8 @@ export function instructionParamsFactory(
       return parseAtaInitInstructions(instructions);
     case TransactionType.StakingAuthorize:
       return parseStakingAuthorizeInstructions(instructions);
+    case TransactionType.StakingAuthorizeRaw:
+      return parseStakingAuthorizeInstructions(instructions);
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/src/lib/instructionParamsFactory.ts:63]: For this parse, shouldn't we also throw if the # of instructions and that each instruction is for an expected program?

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]

File: modules/sdk-coin-sol/src/lib/transaction.ts (hunk 8 above)
Comment (zahin-mohammad) [reply modules/sdk-coin-sol/src/lib/transaction.ts:481]: Also seems you might be missing tests for `explainTransaction` in this scenario? Maybe i missed them.

File: modules/sdk-coin-sol/src/sol.ts (hunk 7 above)
Comment (ewangbitgo) [reply modules/sdk-coin-sol/src/sol.ts:420]: yeah, compiler still complain if not.

File: modules/sdk-coin-sol/src/lib/instructionParamsFactory.ts (hunk 9 above)
Comment (ewangbitgo) [reply modules/sdk-coin-sol/src/lib/instructionParamsFactory.ts:63]: good catch, this should be removed as I moved the new builder away from TransactionBuilder but BaseTransactionBuilder

File: modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts (hunk 4 above)
Comment (ewangbitgo) [reply modules/sdk-coin-sol/src/lib/stakingRawMsgAuthorizeBuilder.ts]: yes, builder.from will call this

File: modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts (hunk 6 above)
Comment (ewangbitgo) [reply modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts:6]: added

File: modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+      tx.inputs.length.should.equal(0);
+      tx.outputs.length.should.equal(0);
+      const rawTx = tx.toBroadcastFormat();
+      const signable = tx.signablePayload.toString('base64');
+      should.equal(signable, testData.STAKING_AUTHORIZE_RAW_MSG);
+      should.equal(Utils.isValidRawTransaction(rawTx), true);
+      should.equal(rawTx, testData.STAKING_AUTHORIZE_RAW_MSG_TXN);
+    });
+
+    it('explain a transaction', async () => {
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts:46]:
```suggestion
    it('should explain a transaction', async () => {
```

File: modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+      });
+      should.deepEqual(explain.stakingAuthorize, {
+        stakingAddress: 'J8cECxcT6Q6H4fcQCvd4LbhmmSjsHL63kpJtrUcrF74Q',
+        oldWithdrawAddress: '6xgesG4vajCYfAQpknodrarD49ZCnXGvYA4H1DLuGV7Y',
+        newWithdrawAddress: '4p1VdN6BngTAbWR7Q5JPpbB6dc4k4y8wn1knmmWEjc9i',
+        custodianAddress: 'DHCVjKy7kN6D6vM69nHcEeEeS685qtonFbiFNBW5bGiq',
+      });
+    });
+
+    it('build from an unsigned transaction', async () => {
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts:34]:
```suggestion
    it('should build from an unsigned transaction', async () => {
```

File: modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts [hunk 12]
```diff
@@ -0,0 +1,90 @@
+import should from 'should';
+import * as testData from '../../resources/sol';
+import { getBuilderFactory } from '../getBuilderFactory';
+import { Utils, Transaction } from '../../../src';
+
+describe('Sol Staking Raw Message Authorize Builder', () => {
+  const factory = getBuilderFactory('tsol');
+
+  describe('Succeed', () => {
+    it('build a create staking authorization unsigned tx', async () => {
```
Comment (zahin-mohammad) [inline modules/sdk-coin-sol/test/unit/transactionBuilder/staingRawMsgAuthorizeBuilder.ts:10]:
```suggestion
    it('should build a create staking authorization unsigned tx', async () => {
```

**Review by zahin-mohammad — DISMISSED** [review]

File: modules/sdk-coin-sol/src/lib/utils.ts (hunk 5 above)
Comment (rlutzbitgo) [reply modules/sdk-coin-sol/src/lib/utils.ts:350]: I see that the constants were defined, but still not used here

File: modules/sdk-coin-sol/src/lib/instructionParamsFactory.ts [hunk 13]
```diff
@@ -563,6 +565,39 @@ function parseStakingAuthorizeInstructions(
...
+    },
+  };
+  instructionData.push(nonce);
+  const authorize = instructions[1];
+  assert(authorize.keys.length === 5, 'Invalid number of keys in authorize instruction');
+  instructionData.push({
+    type: InstructionBuilderTypes.StakingAuthorize,
+    params: {
+      stakingAddress: authorize.keys[0].pubkey.toString(),
+      oldAuthorizeAddress: authorize.keys[2].pubkey.toString(),
```
Comment (rlutzbitgo) [inline modules/sdk-coin-sol/src/lib/instructionParamsFactory.ts:593]: What is authorize.keys[1]?

**Review by rlutzbitgo — APPROVED** [review]

**Review by zahin-mohammad — APPROVED** [review]

**Pull request merged**